rfcs/51-animation-composition.md

```diff
@@ -0,0 +1,450 @@
+other clips in the graph, it's influence on the final output grows. These weights
+may evolve over time to allow a smooth transition between animations.
+
+A common example of this kind of blending is smoothing out a transition from one
+animation clip to another. As a character stops walking and starts running, bones
+do not immediately snap into place when starting a dash. Blending can help here
+by smoothing out the transition. Upon beginning the transition, the blend is
+weighted only towards the walk, but smoothly transitions over time to favor the
+run. This removes any potential discontinuities and provides for a visually
+appealing transition from one animation to another.
+
+Another example allows developers to convey a range of game states to the player
+via animation. For example, if a character is hurt, running may get progressively
+harder and harder. To convey this to the player, developers may decrease the
+overall movement speed of the character while blending the normal walk and run
+animations with a limp animation. As the character gets increasingly damaged, the
+blend becomes more and more weighted towards the limp animation. Unlike the
+transition case, this is a persistent blend and is not driven by time but by
+gameplay design.
+
+Each clip has its on assigned time. Some clips can advance faster or slower than
+others. This allows more fine grained control over individual animations than
+using the global time scaling feature to speed up or slow down animations. This
+can be tied to some gameplay element to produce more believable effects. For
+example, speeding up the run animation for a character when they pick up a speed
+up can be a cheap and effective way to convey the power up without needing to
+author a speed up manually.
+
+### Graph Nodes
+To help manage the blending of a large number of clips, an animation graph is
+comprised of a network of connected graph nodes. A public facing API allows
+developers to create, get, and mutate nodes. Nodes can be of either two types:
+blend node or clip node. The graph always has one single root node of either
+type.
+
+Clip nodes cannot have inputs, but contain a reference to an animation clip. A
+animation graph can have multiple clip nodes referring to same underlying clip.
+
+Blend nodes mix all of the provided inputs using a weighted blend. Blend nodes do
+not contain any animation of their own, but can have zero or more inputs. Each of
+these edges has a `f32` weight associated with it and points to a source node,
+which can be any other node, including another blend node. The weight
+of parent to child edge affects the end weighting of all descendant clip nodes.
+
+Every node has it's own time value. For clip nodes, this is the time at which the
+underlying clip is sampled at. For blend nodes, this value doesn't do anything.
+However, an optional setting will allow a blend node to propagate the time value
```

Comment:
The first common use case that I can think of is "any-time-after" transitions. Where a transition from a source animation is triggered automatically after a fixed time point has passed. That source animation will continue from the same time it had before, but the new animation it's transitioning to will start at zero. Both will advance at their own respective time scales while the transition blends smoothly from the source to the destination. While this is notably very low level, this can be used to build smooth transitions for a state machine like flow at a higher level.

It should be absolute time relative to the underlying clip. The delta is scaled based on the global, graph level, and clip level time scales.

Reply:
> For the propagation, this is largely just a QoL feature for letting you advance an entire subgraph at once.. 

Yeah I think the propagation part might be too "high level" and opinionated at this stage in the game. Useful for some scenarios and not others, "invalid" in some cases (scrubbing results in going "outside" the duration bounds of one animation but not the other, animations are configured to progress at different rates / times, etc). And the behavior (when needed) can be implemented reasonably easily using existing features (ex: `for_each_child_node(parent_node, |child| child.set_time(time)`)

Reply:
(or `for_each_descendant_node`)

Reply:
When reading the preceding paragraph, I had this question: "This seems to try to enable composition, but why a graph system would improve on a simple weight-based approach?"

And this sentence comes in. It gives a feature that is impossible to accomplish without graphs: namely: putting a bunch of animations together and having the capacity to control them exactly like if they were only a single animation.

In this context, I think having time control on blend nodes is really useful.